numpy/core/shape_base.py

```diff
@@ -609,12 +740,65 @@ def block(arrays):
 
 
     """
-    bottom_index, arr_ndim = _block_check_depths_match(arrays)
+    arrays, list_ndim, result_ndim, final_size = _block_setup(arrays)
+
+    # It was found through benchmarking that making an array of final size
+    # around 256x256 was faster by straight concatenation on a
+    # i7-7700HQ processor and dual channel ram 2400MHz.
+    # It didn't seem to matter heavily on the dtype used.
+    #
+    # A 2D array using repeated concatenation requires 2 copies of the array.
+    #
+    # The fastest algorithm will depend on the ratio of CPU power to memory
+    # speed.
+    # One can monitor the results of the benchmark
+    # https://pv.github.io/numpy-bench/#bench_shape_base.Block2D.time_block2d
+    # to tune this parameter until a C version of the `_block_info_recusion`
+    # algorithm is implemented which would likely be faster than the python
+    # version.
+    if list_ndim * final_size > (2 * 512 * 512):
+        return _block_slicing(arrays, list_ndim, result_ndim)
+    else:
+        return _block_concatenate(arrays, list_ndim, result_ndim)
+
+
+# Theses helper functions are mostly used for testing.
+# They allow us to write tests that directly call `_block_slicing`
+# or `_block_concatenate` wtihout blocking large arrays to forse the wisdom
+# to trigger the desired path.
+def _block_setup(arrays):
+    """
+    Returns
+    (`arrays`, list_ndim, result_ndim, final_size)
+    """
+    bottom_index, arr_ndim, final_size = _block_check_depths_match(arrays)
     list_ndim = len(bottom_index)
     if bottom_index and bottom_index[-1] is None:
         raise ValueError(
             'List at {} cannot be empty'.format(
                 _block_format_index(bottom_index)
             )
         )
-    return _block(arrays, list_ndim, max(arr_ndim, list_ndim))
+    result_ndim = max(arr_ndim, list_ndim)
+    return arrays, list_ndim, result_ndim, final_size
+
+
+def _block_slicing(arrays, list_ndim, result_ndim):
+    shape, slices, arrays = _block_info_recursion(
+        arrays, list_ndim, result_ndim)
+    dtype = _nx.result_type(*[arr.dtype for arr in arrays])
+    result = _nx.empty(shape=shape, dtype=dtype)
```

Comment:
I was just looking at numpy's concatenate code, which led me to notice this doesn't account for memory order. That is, the current `np.block` (using concatenate internally) preserves the memory order from the input:
```python
>>> N, M = 256, 16
>>> np.block( [[np.ones((N, N))]*M]*M).strides
(32768, 8)
>>> np.block( [[np.ones((N, N), order='F')]*M]*M).strides
(8, 32768)
```
This PR's new implementation does not preserve the memory order, it always outputs "C" order.

But I don't think it matters. The docstring of `np.block` doesn't say anything about preserving order, so my feeling is it is fine to change it. I think the main reason `np.concatenate` preserves order was that it gives a speedup for non-C-order arrays, but as far as I tested the speedup we now get from fewer copies far outweighs any speedup from memory order. Most people use C-order arrays anyway.

Also, the `np.concatenate` code does a complex calculation in C code in order to get the optimal output memory order. The calculation may be quite difficult to reasonably implement in python *.


I just wanted to bring it up. But I think we can ignore it.

<details>
<summary>*</summary>
Completely untested attempt at a python implementation of `PyArray_CreateMultiSortedStridePerm` you would need here:

```python
strides = zip(*(a.strides for a in arrays))
stride_cmp = lambda a, b: (1 if all(abs(ai) < abs(bi) for ai,bi in zip(a[1],b[1])) else -1)
order = next(zip(*sorted(enumerate(strides), key=functools.cmp_to_key(stride_cmp))))
result = np.empty(np.array(shape)[order], dtype=dtype).transpose(*order)
```
</details>

Reply:
@ahaldane I think C/F order can get quite annoying if we flip a whole set of 'F' arrays to 'C'. It might be reasonable to keep a counter of the '`C' elements, and of the 'F' elements. How about

```python
    C_orders = sum(arr.size for arr in arrays if arr.flags['C_CONTIGUOUS'])
    F_orders = sum(arr.size for arr in arrays if arr.flags['F_CONTIGUOUS'])
    order = 'C' if C_order >= F_orders else 'F'
    result = _nx.empty(shape=shape, dtype=dtype, order=order)
```

Reply:
At least with that, it will preserve the memory order in the two extremes (All F) (All C) and default to C when the counts are equal.

```
In [13]: N,M = 64,16

In [14]:  np.block( [[[np.ones((N, N), order='F')]*M]*M]).shape
Out[14]: (1, 1024, 1024)
```

Reply:
Hmm I like that idea. We still wouldn't guarantee anything about the output order, but at least we would speedup the case of all-fortran arrays.

Instead of depending on the size, though, I would only use 'F' order output if every input array satisfies `np.isfortran(a)`. Otherwise, if there is even a single non-F-array, use C order.

For all-fortran arrays, it looks like I get a 30% speedup by doing that (tested for `N,M = 256,16`), so seems worth it. I get no observable slowdown for the C-order case.

Reply:
Actually I don't think that's quite right. The F-order test can be a little tricky because 1d arrays count as both C- and F-ordered. We somehow need to make sure the arrays are all F-ordered and *not* C-ordered.

Reply:
And these are the results of your tested algorithm

```python
    strides = zip(*(a.strides for a in arrays))
    def stride_cmp(a, b):
        if all(abs(ai) < abs(bi) for ai,bi in zip(a[1],b[1])):
            return -1
        else:
            return 1

    order = next(zip(*sorted(enumerate(strides), key=functools.cmp_to_key(stride_cmp))))
    result = _nx.empty(_nx.array(shape)[_nx.array(order)],
                       dtype=dtype).transpose(*order)
```

```python
In [11]: N,M = 64,16

In [12]: np.block( [[[np.ones((N, N), order='F')]*M]*M]).strides
Out[12]: (8388608, 8192, 8)

In [13]: np.block( [[[np.ones((N, N), order='C')]*M]*M]).strides
Out[13]: (8, 8, 8192)
```

does functools.cmp_to_key make this an O(N^2) sorting algorithm?

Reply:
This might do it:
```python    
    F_order = all(arr.flags['F_CONTIGUOUS'] for arr in arrays)
    C_order =  all(arr.flags['C_CONTIGUOUS'] for arr in arrays)
    order = 'F' if F_order and not C_order else 'C'
    result = _nx.empty(shape=shape, dtype=dtype, order=order)
```

Reply:
Would it even be possible to export the `PyArray_CreateMultiSortedStridePerm` API and leave it as "internal only"

Reply:
I found our last discussion of it [here](https://github.com/numpy/numpy/pull/9209#discussion_r138170238), and the "bug" referred to is #2696. I guess there was no inherent problem with the function, so maybe it's ok to expose in python.

Reply:
This is really funny, because I might soon be hacking my code to use a non-standard ordering, so I might be interested in having some automagic order detection. I'll think about it.

Reply:
If you do end up trying exporting it, I would put the python func in `numpy.lib.stride_tricks`, as a separate PR. It might be a nice addition there!

Reply:
Also, by the way, as I was reading this PR, I was making notes for a C port of it, so maybe we can port to C soon anyway. Then porting the ordering code to python is unnecessary effort...

Reply:
I'm mostly interested in preserving the order when all matricies have the same order (modulo ambiguitiy, probably hard to detect :/ correctly). 
That is the case I want to ensure catching. 
I'm not happy with the results of `concatenate` you showed. I want to make sure to read the previous discussion you linked to before opening a new issue for this.

I'm also unhappy with how the order of the resulting matrices might be different between the two algorithms. Part of me doesn't care since small matrices are probably easier to copy than to ask python to think about them.